packages/ra-core/src/types.ts

```diff
@@ -62,3 +66,32 @@ export interface ReduxState {
 export type Dispatch<T> = T extends (...args: infer A) => any
     ? (...args: A) => void
     : never;
+
+type RenderResourcesFunction = (permissions: any) => any[];
+export type AdminChildren = RenderResourcesFunction | ReactNode;
```

Comment:
ReactNode is a bit too broad here, we could use the Resource props and impose a `ReactElement<ResourceProps>`